File: trees/leaf-count.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node * left;
    struct node * right;
};

int leaf_count(struct node * node) {
    if (node == NULL)
        return 0;

    if (node->left == NULL && node->right == NULL)
        return 1;
    
    return leaf_count(node->left) 
         + leaf_count(node->right);
}

struct node * new(int data) {
    struct node * node = (struct node *) malloc(sizeof(struct node));

    node->data = data;
    node->left = NULL;
    node->right = NULL;

    return node;
}

int main() {
    struct node * root = new(1);
    root->left = new(2);
    root->right = new(3);
    root->left->left = new(4);
    root->left->right = new(5);

    printf("Leaf count of tree is: %d\n", leaf_count(root));

    getchar();
}

File: trees/leaf-count.py
class TreeNode(object):
    def __init__(self, data=None, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    @classmethod
    def leaf_count(cls, node):
        if not node:
            return 0

        if node.left is None and node.right is None:
            return 1

        return cls.leaf_count(node.left) \
             + cls.leaf_count(node.right)


if __name__ == "__main__":
    node = TreeNode(1,
               left=TreeNode(2,
                   left=TreeNode(3),
                   right=TreeNode(5)),
               right=TreeNode(4))

    print "Leaf count of tree is: {}".format(TreeNode.leaf_count(node))

